File: Day020_2025/Q020_Leetcode.py
from typing import List
class Solution:
    def firstCompleteIndex(self, arr: List[int], mat: List[List[int]]) -> int:
        # Map to store the index of each element in the array `arr`
        num_to_index = {num: index for index, num in enumerate(arr)}
        answer = float('inf')
        num_rows = len(mat)
        num_cols = len(mat[0])

        # Check each row for the first complete index
        for row in range(num_rows):
            last_element_index = float('-inf')
            for col in range(num_cols):
                index_val = num_to_index[mat[row][col]]
                last_element_index = max(last_element_index, index_val)
            answer = min(answer, last_element_index)

        # Check each column for the first complete index
        for col in range(num_cols):
            last_element_index = float('-inf')
            for row in range(num_rows):
                index_val = num_to_index[mat[row][col]]
                last_element_index = max(last_element_index, index_val)
            answer = min(answer, last_element_index)

        return answer

File: Day020_2025/Q020_Leetcode.java
import java.util.*;

class Q020_Leetcode {
    public int firstCompleteIndex(int[] arr, int[][] mat) {
        // Map to store the index of each element in the array `arr`
        Map<Integer, Integer> numToIndexMap = new HashMap<>();
        for (int index = 0; index < arr.length; index++) {
            numToIndexMap.put(arr[index], index);
        }

        int answer = Integer.MAX_VALUE;
        int numRows = mat.length;
        int numCols = mat[0].length;

        // Check each row for the first complete index
        for (int row = 0; row < numRows; row++) {
            int lastElementIndex = Integer.MIN_VALUE;
            for (int col = 0; col < numCols; col++) {
                int indexVal = numToIndexMap.get(mat[row][col]);
                lastElementIndex = Math.max(lastElementIndex, indexVal);
            }
            answer = Math.min(answer, lastElementIndex);
        }

        // Check each column for the first complete index
        for (int col = 0; col < numCols; col++) {
            int lastElementIndex = Integer.MIN_VALUE;
            for (int row = 0; row < numRows; row++) {
                int indexVal = numToIndexMap.get(mat[row][col]);
                lastElementIndex = Math.max(lastElementIndex, indexVal);
            }
            answer = Math.min(answer, lastElementIndex);
        }

        return answer;
    }
}
